Rustc shows unnecessary big error when glob reexport failed

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
pub const ITEM: Item = Item;

pub struct Item;

pub fn item() {}

// Glob (*) is required to cause the error
pub use doesnt_exist::*;

mod a {
    use crate::{item, Item, ITEM};
}

mod b {
    use crate::item;
    use crate::Item;
    use crate::ITEM;
}

mod c {
    use crate::item;
}
```
([play](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=d7d38220140a6793756d30e20fdbdc22))


I expected to see an error that only says that "`doesnt_exist` is unresolved".

But the error says that all uses of items from the module with glob import (`crate` in this case) are unresolved too:

```text
error[E0432]: unresolved imports `doesnt_exist`, `crate::item`, `crate::Item`, `crate::ITEM`, `crate::item`, `crate::Item`, `crate::ITEM`, `crate::item`
  --> src/lib.rs:41:9
   |
41 | pub use doesnt_exist::*;
   |         ^^^^^^^^^^^^ use of undeclared crate or module `doesnt_exist`
...
44 |     use crate::{item, Item, ITEM};
   |                 ^^^^  ^^^^  ^^^^
...
48 |     use crate::item;
   |         ^^^^^^^^^^^
49 |     use crate::Item;
   |         ^^^^^^^^^^^
50 |     use crate::ITEM;
   |         ^^^^^^^^^^^
...
54 |     use crate::item;
   |         ^^^^^^^^^^^
```

This may lead to extremely long and unreadable errors. The issue was originally found while developing `teloxide` where it caused [1500+ lines error](https://gist.github.com/WaffleLapkin/0283b070e38fed0affdf5b787b0010d3).

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.51.0-nightly (c97f11af7 2021-01-10)
binary: rustc
commit-hash: c97f11af7bc4a6d3578f6a953be04ab2449a5728
commit-date: 2021-01-10
host: x86_64-unknown-linux-gnu
release: 1.51.0-nightly
```

<details>
<summary>Output on 1.38:</summary>

```
error[E0432]: unresolved import `doesnt_exist`
 --> <source>:8:9
  |
8 | pub use doesnt_exist::*;
  |         ^^^^^^^^^^^^ maybe a missing crate `doesnt_exist`?

warning: unused imports: `ITEM`, `Item`, `item`
  --> <source>:11:17
   |
11 |     use crate::{item, Item, ITEM};
   |                 ^^^^  ^^^^  ^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: unused import: `crate::item`
  --> <source>:15:9
   |
15 |     use crate::item;
   |         ^^^^^^^^^^^

warning: unused import: `crate::Item`
  --> <source>:16:9
   |
16 |     use crate::Item;
   |         ^^^^^^^^^^^

warning: unused import: `crate::ITEM`
  --> <source>:17:9
   |
17 |     use crate::ITEM;
   |         ^^^^^^^^^^^

warning: unused import: `crate::item`
  --> <source>:21:9
   |
21 |     use crate::item;
   |         ^^^^^^^^^^^

error: aborting due to previous error
```
</details>

@rustbot label A-diagnostics A-resolve regression-from-stable-to-stable

--- a/compiler/rustc_resolve/src/imports.rs
+++ b/compiler/rustc_resolve/src/imports.rs
@@ -475,15 +475,15 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
         let indeterminate_imports = mem::take(&mut self.indeterminate_imports);
 
         for (is_indeterminate, import) in determined_imports
-            .into_iter()
+            .iter()
             .map(|i| (false, i))
-            .chain(indeterminate_imports.into_iter().map(|i| (true, i)))
+            .chain(indeterminate_imports.iter().map(|i| (true, i)))
         {
-            let unresolved_import_error = self.finalize_import(import);
+            let unresolved_import_error = self.finalize_import(*import);
 
             // If this import is unresolved then create a dummy import
             // resolution for it so that later resolve stages won't complain.
-            self.import_dummy_binding(import, is_indeterminate);
+            self.import_dummy_binding(*import, is_indeterminate);
 
             if let Some(err) = unresolved_import_error {
                 if let ImportKind::Single { source, ref source_bindings, .. } = import.kind {
@@ -505,27 +505,34 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
                     errors = vec![];
                 }
                 if seen_spans.insert(err.span) {
-                    errors.push((import, err));
+                    errors.push((*import, err));
                     prev_root_id = import.root_id;
                 }
-            } else if is_indeterminate {
-                let path = import_path_to_string(
-                    &import.module_path.iter().map(|seg| seg.ident).collect::<Vec<_>>(),
-                    &import.kind,
-                    import.span,
-                );
-                let err = UnresolvedImportError {
-                    span: import.span,
-                    label: None,
-                    note: None,
-                    suggestion: None,
-                    candidates: None,
-                };
-                // FIXME: there should be a better way of doing this than
-                // formatting this as a string then checking for `::`
-                if path.contains("::") {
-                    errors.push((import, err))
-                }
+            }
+        }
+
+        if !errors.is_empty() {
+            self.throw_unresolved_import_error(errors);
+            return;
+        }
+
+        for import in &indeterminate_imports {
+            let path = import_path_to_string(
+                &import.module_path.iter().map(|seg| seg.ident).collect::<Vec<_>>(),
+                &import.kind,
+                import.span,
+            );
+            let err = UnresolvedImportError {
+                span: import.span,
+                label: None,
+                note: None,
+                suggestion: None,
+                candidates: None,
+            };
+            // FIXME: there should be a better way of doing this than
+            // formatting this as a string then checking for `::`
+            if path.contains("::") {
+                errors.push((*import, err))
             }
         }
 

--- a/tests/ui/imports/issue-81413.rs
+++ b/tests/ui/imports/issue-81413.rs
@@ -1,0 +1,23 @@
+pub const ITEM: Item = Item;
+
+pub struct Item;
+
+pub fn item() {}
+
+pub use doesnt_exist::*;
+//~^ ERROR unresolved import `doesnt_exist`
+mod a {
+    use crate::{item, Item, ITEM};
+}
+
+mod b {
+    use crate::item;
+    use crate::Item;
+    use crate::ITEM;
+}
+
+mod c {
+    use crate::item;
+}
+
+fn main() {}

--- a/tests/ui/imports/issue-81413.stderr
+++ b/tests/ui/imports/issue-81413.stderr
@@ -1,0 +1,11 @@
+error[E0432]: unresolved import `doesnt_exist`
+  --> $DIR/issue-81413.rs:7:9
+   |
+LL | pub use doesnt_exist::*;
+   |         ^^^^^^^^^^^^ maybe a missing crate `doesnt_exist`?
+   |
+   = help: consider adding `extern crate doesnt_exist` to use the `doesnt_exist` crate
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0432`.

